coreax/solvers/recombination.py

```diff
@@ -0,0 +1,609 @@
+    indices *= non_zero_weights_mask
+    indices += zero_weights_mask * max_index
+    # Renormalize the maximum weight; ensures the weight sum is preserved under the new
+    # (co-linearized) indices; prevents co-linearization from changing the weight sum.
+    weights = weights.at[max_index].divide(n_zeros + 1)
+    return nodes[indices], weights[indices], indices
+
+
+# pylint: disable=line-too-long
+# Credit: https://github.com/patrick-kidger/lineax/blob/9b923c8df6556551fedc7adeea7979b5c7b3ffb0/lineax/_solver/svd.py#L67  # noqa: E501
+# for the rank determination code.
+# pylint: enable=line-too-long
+def _resolve_null_basis(
+    nodes: Shaped[Array, "n m"],
+    rcond: Union[float, None] = None,
+) -> tuple[Shaped[Array, "n n"], int]:
+    r"""
+    Resolve the largest left null space basis, and its rank, for passed the node matrix.
+
+    By largest left null space basis, we mean the null space basis under the assumption
+    that the rank of the null space is maximal (assumed to be ``n``). If the rank is not
+    maximal, then only the first :math:`n - m^\prime` basis vectors will be actual basis
+    vectors for the null space (where :math:`m^\prime` is the rank of the node matrix).
+    The remaining "basis" vectors can, and should, be ignored in upstream computations
+    by using the left null space rank value as a cut-off index.
+
+    :param nodes: Matrix of nodes (m-vectors) whose null space is to be determined
+    :param rcond: The relative condition number of the Matrix of nodes
+    :return: The largest left null space basis and its rank, for the passed node matrix.
+    """
+    q, s, _ = jsp.linalg.svd(nodes, full_matrices=True)
+    rcond = _resolve_rcond(nodes.shape, s.dtype, rcond)
+    if s.size > 0:
+        rcond *= jnp.max(s[0])
+    mask = s > rcond
+    matrix_rank = sum(mask)
+    null_space_rank = jnp.maximum(0, nodes.shape[0] - matrix_rank)
+    largest_null_space_basis = q.T[::-1]
+    return largest_null_space_basis, null_space_rank
+
+
+# pylint: disable=line-too-long
+# Credit: https://github.com/patrick-kidger/lineax/blob/9b923c8df6556551fedc7adeea7979b5c7b3ffb0/lineax/_misc.py#L34  # noqa: E501
+# pylint: enable=line-too-long
+def _resolve_rcond(
+    shape: tuple[int, ...], dtype: DTypeLike, rcond: Optional[float] = None
+) -> float:
+    """
+    Resolve the relative condition number (rcond).
+
+    :param shape: The shape of the matrix whose relative condition number to resolved
+    :param dtype: The element dtype of the matrix whose rcond is to be resolved
+    :param rcond: The relative condition number of a given matrix; if ``None``,
+        ``rcond = dtype_floating_point_eps * max(shape)``; else if negative,
+        ``rcond = dtype_floating_point_eps``
+    :return: The resolved relative condition number (rcond)
+    """
+    if rcond is None:
+        return jnp.finfo(dtype).eps * max(shape)
+    return jnp.where(rcond < jnp.asarray(0), jnp.finfo(dtype).eps, rcond)
```

Comment:
Move to `util.py`?

Reply:
IF so, I'd like to a unit test for this as well

Reply:
Or if not I think an issue should be made to move it and write a unit test for it

Reply:
This function is "private" here as its implementation is solely for achiving the recombination algorithms. I think making this into a proper utillity function should be left for a separate issue, considering the pull is large already.

Reply:
I agree, could you make that issue?

Reply:
See issue #769